src/webgpu/util/f32_interval.ts

```diff
@@ -0,0 +1,629 @@
+  } else {
+    // x_32 rounded towards -inf, so is before x
+    const other = nextAfter(x_32, true, false).value as number;
+    return [converted, other];
+  }
+}
+
+/** Convert a point to an interval of valid rounded values, both flushed and not flushed */
+function roundAndFlushPointToInterval(n: number, fn: PointToInterval) {
+  assert(!Number.isNaN(n), `flush not defined for NaN`);
+  const values = roundedF32Values(n);
+  const inputs = new Set<number>([...values, ...values.map(flushSubnormalNumber)]);
+  const results = new Set<F32Interval>([...inputs].map(fn));
+  return F32Interval.span(...results);
+}
+
+/** Convert a pair of points to an interval of valid rounded values, both flushed and not flushed */
+function roundAndFlushBinaryToInterval(x: number, y: number, fn: BinaryToInterval): F32Interval {
+  assert(!Number.isNaN(x), `flush not defined for NaN`);
+  assert(!Number.isNaN(y), `flush not defined for NaN`);
+  const x_values = roundedF32Values(x);
+  const y_values = roundedF32Values(y);
+  const x_inputs = new Set<number>([...x_values, ...x_values.map(flushSubnormalNumber)]);
+  const y_inputs = new Set<number>([...y_values, ...y_values.map(flushSubnormalNumber)]);
+  const intervals = new Set<F32Interval>();
+  x_inputs.forEach(inner_x => {
+    y_inputs.forEach(inner_y => {
+      intervals.add(fn(inner_x, inner_y));
+    });
+  });
+  return F32Interval.span(...intervals);
+}
+
+/** Convert a triplet of points to an interval of valid rounded values, both flushed and not flushed */
+function roundAndFlushTernaryToInterval(
+  x: number,
+  y: number,
+  z: number,
+  fn: TernaryToInterval
+): F32Interval {
+  assert(!Number.isNaN(x), `flush not defined for NaN`);
+  assert(!Number.isNaN(y), `flush not defined for NaN`);
+  assert(!Number.isNaN(z), `flush not defined for NaN`);
+  const x_values = roundedF32Values(x);
+  const y_values = roundedF32Values(y);
+  const z_values = roundedF32Values(z);
+  const x_inputs = new Set<number>([...x_values, ...x_values.map(flushSubnormalNumber)]);
+  const y_inputs = new Set<number>([...y_values, ...y_values.map(flushSubnormalNumber)]);
+  const z_inputs = new Set<number>([...z_values, ...z_values.map(flushSubnormalNumber)]);
+  const intervals = new Set<F32Interval>();
+  // prettier-ignore
+  x_inputs.forEach(inner_x => {
+    y_inputs.forEach(inner_y => {
+      z_inputs.forEach(inner_z => {
+        intervals.add(fn(inner_x, inner_y, inner_z));
+      });
+    });
+  });
+
+  return F32Interval.span(...intervals);
+}
+
+/** Convert a point to an interval correctly rounded around the point */
+export function correctlyRoundedInterval(n: number): F32Interval {
+  return roundAndFlushPointToInterval(n, (n: number) => {
+    assert(!Number.isNaN(n), `absolute not defined for NaN`);
+    return toInterval(n);
+  });
+}
+
+/** Convert a point to an interval of absolute error around the point */
+export function absoluteErrorInterval(n: number, rng: number): F32Interval {
```

Comment:
nit: Please explain the 'rng'.  I don't understand the mnemonic.
It's the absolute error.

Reply:
Will do when productionizing